File: leetcode/567_permutation_in_string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        unordered_map<char, int> freq;
        for (char& c : s1) {
            freq[c]++;
        }
        int len = s2.length();
        int windowStart = 0;
        int windowEnd = 0;
        while (windowEnd < len) {
            if (freq.find(s2[windowEnd]) != freq.end() && freq[s2[windowEnd]] > 0) {
                freq[s2[windowEnd]]--;
                windowEnd++;
                if (windowEnd - windowStart == s1.length()) {
                    return true;
                }
            } else if (windowStart == windowEnd) {
                windowStart++;
                windowEnd++;
            } else {
                freq[s2[windowStart]]++;
                windowStart++;
            }
        }
        return false;
    }
};

File: leetcode/567_permutation_in_string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        occurrences = {}
        for c in s1:
            if c not in occurrences:
                occurrences[c] = 1
            else:
                occurrences[c] += 1
        l = 0
        r = 0
        n = len(s2)
        while r < n:
            counter = 0
            while r < n and occurrences.get(s2[r], 0) > 0:
                occurrences[s2[r]] -= 1
                counter += 1
                r += 1
            if counter == len(s1):
                return True
            if r >= n:
                break
            if l == r:
                r += 1
                l += 1
            else:
                occurrences[s2[l]] += 1
                l += 1
        if (r-l == len(s1)):
            return True
        return False
